cloud/pkg/admissioncontroller/admit_device.go

```diff
@@ -0,0 +1,68 @@
+package admissioncontroller
+
+import (
+	"net/http"
+	"reflect"
+	"strings"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog"
+
+	devicesv1alpha1 "github.com/kubeedge/kubeedge/cloud/pkg/apis/devices/v1alpha1"
+)
+
+func admitDevice(review admissionv1beta1.AdmissionReview) *admissionv1beta1.AdmissionResponse {
+	reviewResponse := admissionv1beta1.AdmissionResponse{}
+	reviewResponse.Allowed = true
+	var msg string
+	switch review.Request.Operation {
+	case admissionv1beta1.Create:
+		raw := review.Request.Object.Raw
+		device := devicesv1alpha1.Device{}
+		deserializer := codecs.UniversalDeserializer()
+		if _, _, err := deserializer.Decode(raw, nil, &device); err != nil {
+			klog.Errorf("validation failed with error: %v", err)
+			return toAdmissionResponse(err)
+		}
+
+		un := unstructured.Unstructured{}
+		if err := un.UnmarshalJSON(raw); err != nil {
+			klog.Errorf("failed to unmarshall object (%v) with error: %v", string(raw), err)
+			return toAdmissionResponse(err)
+		}
+
+		mdSpec, ok := un.Object["spec"].(map[string]interface{})
+		if ok {
+			mdProto, ok := mdSpec["protocol"].(map[string]interface{})
+			if ok {
+				keys := reflect.ValueOf(mdProto).MapKeys()
+				//device should only have one protocol defined.
+				if len(keys) > 1 {
+					reviewResponse.Allowed = false
+					msg = msg + "multiple protocols found!"
```

Comment:
nit: please add a blank space either before or after the appended message to make it more readable

Reply:
Good catch!